processing/src/main/java/io/druid/segment/virtual/ExpressionObjectSelector.java

```diff
@@ -83,10 +83,7 @@ public Number get()
         // Unknown ValueType. Try making an Object selector and see if that gives us anything useful.
         final ObjectColumnSelector selector = columnSelectorFactory.makeObjectColumnSelector(columnName);
         final Class clazz = selector == null ? null : selector.classOfObject();
-        if (selector == null || (clazz != Object.class && Number.class.isAssignableFrom(clazz))) {
-          // We know there are no numbers here. Use a null supplier.
-          supplier = null;
-        } else {
+        if (selector != null && (clazz == Object.class || Number.class.isAssignableFrom(clazz))) {
```

Comment:
Sure I'll add a test.

`clazz.isAssignableFrom(Number.class)` is not exactly what I'm trying to go for, though. This code path might also be parsing Strings (see tryParse). I guess this could be `clazz.isAssignableFrom(Number.class) || clazz.isAssignableFrom(String.class)`.